src/box/lua/schema.lua

```diff
@@ -3970,12 +4046,69 @@ local function drop(uid, level)
     for _, tuple in pairs(privs) do
         -- we need an additional box.session.su() here, because of
         -- unnecessary check for privilege PRIV_REVOKE in priv_def_check()
-        box.session.su("admin", revoke, level + 2, uid, uid, tuple.privilege,
-                       tuple.object_type, tuple.object_id)
+        if tuple[PRIV_OPTS_FIELD_ID] ~= nil and
+            tuple[PRIV_OPTS_FIELD_ID].origins ~= nil then
+            for origin, priv in pairs(tuple[PRIV_OPTS_FIELD_ID].origins) do
+                box.session.su('admin', revoke, level + 2, uid, uid, priv,
+                               tuple.object_type, tuple.object_id,
+                               {_origin = origin})
+            end
+        else
+            box.session.su('admin', revoke, level + 2, uid, uid,
+                           tuple.privilege, tuple.object_type, tuple.object_id,
+                           {_origin = DEFAULT_ORIGIN})
+        end
     end
     box.space[box.schema.USER_ID]:delete{uid}
 end
 
+local function drop(uid, level, opts)
+    opts = opts or {}
+    local origin = opts._origin or DEFAULT_ORIGIN
+    local tuple = box.space[box.schema.USER_ID]:get{uid}
+    local origins = user_origins_from_tuple(tuple)
+    if origins[origin] == nil then
+        if opts.if_exists then
+            return
+        end
+        local err
+        if tuple.type == 'role' then
+            err = box.error.NO_SUCH_ROLE
+        else
+            err = box.error.NO_SUCH_USER
+        end
+        box.error(err, tuple.name, level + 1)
+    end
+    origins[origin] = nil
+    if next(origins) == nil then
+        full_drop(uid, level)
+        return
+    end
+    local _vpriv = box.space[box.schema.VPRIV_ID]
+    for _, t in pairs(_vpriv.index.primary:select{uid}) do
+        local priv = 0
+        if t[PRIV_OPTS_FIELD_ID] ~= nil
+            and t[PRIV_OPTS_FIELD_ID].origins ~= nil then
+            priv = t[PRIV_OPTS_FIELD_ID].origins[origin] or 0
+        elseif origin == DEFAULT_ORIGIN then
+            priv = t.privilege
+        end
+        if priv ~= 0 then
+            box.session.su('admin', revoke, level + 2, uid, uid, priv,
```

Comment:
Nit: `level + 1` above in this function, but `level + 2` here. Should be the same?

Reply:
Forgot about `box.session.su`, agree.